src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs

```diff
@@ -32,29 +37,31 @@ public ResponseCachingMiddleware(RequestDelegate next, IResponseCache cache)
         public async Task Invoke(HttpContext context)
         {
             var cachingContext = new ResponseCachingContext(context, _cache);
+
             // Should we attempt any caching logic?
-            if (cachingContext.CheckRequestAllowsCaching())
+            if (cachingContext.RequestIsCacheable())
             {
                 // Can this request be served from cache?
                 if (await cachingContext.TryServeFromCacheAsync())
                 {
                     return;
                 }
 
-                // Hook up to listen to the response stream
-                cachingContext.HookResponseStream();
-
                 try
                 {
+                    // Hook up to listen to the response stream
+                    cachingContext.HookResponseStream();
```

Comment:
why was this moved inside the try/finally? If it fails throws what will UnhookResponseStream do?